x/staking/keeper/msg_server.go

```diff
@@ -605,6 +608,86 @@ func (k msgServer) UpdateParams(ctx context.Context, msg *types.MsgUpdateParams)
 	return &types.MsgUpdateParamsResponse{}, nil
 }
 
-func (k msgServer) RotateConsPubKey(_ context.Context, _ *types.MsgRotateConsPubKey) (*types.MsgRotateConsPubKeyResponse, error) {
-	return nil, nil
+func (k msgServer) RotateConsPubKey(ctx context.Context, msg *types.MsgRotateConsPubKey) (res *types.MsgRotateConsPubKeyResponse, err error) {
+	cv := msg.NewPubkey.GetCachedValue()
+	pk, ok := cv.(cryptotypes.PubKey)
+	if !ok {
+		return nil, errorsmod.Wrapf(sdkerrors.ErrInvalidType, "expecting cryptotypes.PubKey, got %T", cv)
+	}
+
+	// check cons key is already present in the key rotation history.
+	rotatedTo, err := k.RotatedConsKeyMapIndex.Get(ctx, pk.Address())
+	if err != nil && !errors.Is(err, collections.ErrNotFound) {
+		return nil, err
+	}
+
+	if rotatedTo != nil {
+		return nil, errorsmod.Wrap(sdkerrors.ErrInvalidAddress,
+			"the address is already present in rotation history, please try with new one")
+	}
+
+	newConsAddr := sdk.ConsAddress(pk.Address())
+
+	// checks if NewPubKey is not duplicated on ValidatorsByConsAddr
+	validator, _ := k.Keeper.ValidatorByConsAddr(ctx, newConsAddr)
+	if validator != nil {
+		return nil, types.ErrConsensusPubKeyAlreadyUsedForValidator
+	}
+
+	valAddr, err := k.validatorAddressCodec.StringToBytes(validator.GetOperator())
+	if err != nil {
+		return nil, err
+	}
+
+	validator, err = k.Keeper.GetValidator(ctx, valAddr)
+	if err != nil {
+		return nil, types.ErrNoValidatorFound
+	}
+
+	if status := validator.GetStatus(); status != types.Bonded {
+		return nil, errorsmod.Wrapf(sdkerrors.ErrInvalidType, "validator status is not bonded, got %s", status.String())
+	}
+
+	// Check if the validator is exceeding parameter MaxConsPubKeyRotations within the
+	// unbonding period by iterating ConsPubKeyRotationHistory.
+	exceedsLimit, err := k.exceedsMaxRotations(ctx, valAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	if exceedsLimit {
+		return nil, types.ErrExceedingMaxConsPubKeyRotations
+	}
```

Comment:
If exceeding limits is always an error scenario then exceedsMaxRotations should simply return an error and not (bool, error).